File: src/main/java/komu/blunt/asm/Label.java
package komu.blunt.asm;

import static com.google.common.base.Preconditions.checkNotNull;

public final class Label {
    
    private final String name;
    private int address = -1;
    
    Label(String name) {
        this.name = checkNotNull(name);
    }

    void setAddress(int address) {
        if (address < 0) throw new IllegalArgumentException("negative address");
        
        if (this.address != -1)
            throw new IllegalStateException("address already set");
        
        this.address = address;
    }
    
    int getAddress() {
        if (address == -1) throw new IllegalStateException("address not initialized");

        return address;
    }

    @Override
    public String toString() {
        return name;
    }
}

File: src/main/kotlin/komu/blunt/asm/Label.kt
package komu.blunt.asm

class Label(private val name: String) {

    fun relocateBy(offset: Int) {
        require(offset >= 0) { "negative offset: $offset" }

        address += offset
    }

    var address: Int = -1
        get() {
            check(field != -1) { "address not initialized" }
            return field
        }

        set(value) {
            require(value >= 0) { "negative address: $value" }
            field = value
        }

    override fun toString() = name
}
